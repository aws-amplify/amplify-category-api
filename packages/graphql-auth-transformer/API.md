## API Report File for "graphql-auth-transformer"

> Do not edit this file. It is a report generated by [API Extractor](https://api-extractor.com/).

```ts

import { CompoundExpressionNode } from 'graphql-mapping-template';
import { default as default_2 } from 'cloudform-types/types/appSync/apiKey';
import { default as default_3 } from 'cloudform-types/types/appSync/resolver';
import { default as default_4 } from 'cloudform-types/types/appSync/dataSource';
import { DirectiveNode } from 'graphql';
import { Expression } from 'graphql-mapping-template';
import { FieldDefinitionNode } from 'graphql';
import GraphQLApi from 'cloudform-types/types/appSync/graphQlApi';
import { InterfaceTypeDefinitionNode } from 'graphql';
import ManagedPolicy from 'cloudform-types/types/iam/managedPolicy';
import { NumberParameter } from 'cloudform-types';
import { ObjectTypeDefinitionNode } from 'graphql';
import { StringParameter } from 'cloudform-types';
import Template from 'cloudform-types/types/template';
import { Transformer as Transformer_2 } from 'graphql-transformer-core';
import { TransformerContext } from 'graphql-transformer-core';

// @public (undocumented)
export type ApiKeyConfig = {
    description?: string;
    apiKeyExpirationDays: number;
    apiKeyExpirationDate?: Date;
};

// @public (undocumented)
export type AppSyncAuthConfiguration = {
    defaultAuthentication: AppSyncAuthConfigurationEntry;
    additionalAuthenticationProviders: Array<AppSyncAuthConfigurationEntry>;
};

// @public (undocumented)
export type AppSyncAuthConfigurationEntry = {
    authenticationType: AppSyncAuthMode;
    apiKeyConfig?: ApiKeyConfig;
    userPoolConfig?: UserPoolConfig;
    openIDConnectConfig?: OpenIDConnectConfig;
    lambdaAuthorizerConfig?: LambdaAuthorizerConfig;
};

// @public (undocumented)
export type AppSyncAuthMode = 'API_KEY' | 'AMAZON_COGNITO_USER_POOLS' | 'AWS_IAM' | 'OPENID_CONNECT' | 'AWS_LAMBDA';

// @public (undocumented)
export type ConfiguredAuthProviders = {
    default: AuthProvider;
    onlyDefaultAuthProviderConfigured: boolean;
    hasApiKey: boolean;
    hasUserPools: boolean;
    hasOIDC: boolean;
    hasIAM: boolean;
};

// @public (undocumented)
export type LambdaAuthorizerConfig = {
    lambdaFunction: string;
    ttlSeconds?: number;
};

// @public (undocumented)
export class ModelAuthTransformer extends Transformer_2 {
    constructor(config?: ModelAuthTransformerConfig);
    // (undocumented)
    after: (ctx: TransformerContext) => void;
    // (undocumented)
    authPolicyResources: Set<string>;
    // (undocumented)
    before: (ctx: TransformerContext) => void;
    // (undocumented)
    config: ModelAuthTransformerConfig;
    // (undocumented)
    configuredAuthProviders: ConfiguredAuthProviders;
    // (undocumented)
    field: (parent: ObjectTypeDefinitionNode | InterfaceTypeDefinitionNode, definition: FieldDefinitionNode, directive: DirectiveNode, ctx: TransformerContext) => void;
    // (undocumented)
    generateIAMPolicyforAuthRole: boolean;
    // (undocumented)
    generateIAMPolicyforUnauthRole: boolean;
    // (undocumented)
    hasProviderAuthRules(rules: AuthRule[]): Boolean;
    // (undocumented)
    object: (def: ObjectTypeDefinitionNode, directive: DirectiveNode, ctx: TransformerContext) => void;
    // Warning: (ae-forgotten-export) The symbol "AuthRule" needs to be exported by the entry point index.d.ts
    //
    // (undocumented)
    protectSyncQuery(ctx: TransformerContext, def: ObjectTypeDefinitionNode, resolverResourceID: string, rules: AuthRule[]): void;
    // Warning: (ae-forgotten-export) The symbol "ResourceFactory" needs to be exported by the entry point index.d.ts
    //
    // (undocumented)
    resources: ResourceFactory;
    // (undocumented)
    unauthPolicyResources: Set<string>;
}

// @public (undocumented)
export type ModelAuthTransformerConfig = {
    authConfig?: AppSyncAuthConfiguration;
    addAwsIamAuthInOutputSchema?: boolean;
};

// @public (undocumented)
export type OpenIDConnectConfig = {
    name: string;
    issuerUrl: string;
    clientId?: string;
    iatTTL?: number;
    authTTL?: number;
};

// @public (undocumented)
export type UserPoolConfig = {
    userPoolId: string;
};

// Warnings were encountered during analysis:
//
// src/ModelAuthTransformer.ts:144:3 - (ae-forgotten-export) The symbol "AuthProvider" needs to be exported by the entry point index.d.ts

// (No @packageDocumentation comment for this package)

```
