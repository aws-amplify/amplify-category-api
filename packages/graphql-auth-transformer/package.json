{
  "name": "graphql-auth-transformer",
  "version": "7.2.74",
  "description": "Implements the @auth directive for the appsync model transform.",
  "repository": {
    "type": "git",
    "url": "https://github.com/aws-amplify/amplify-category-api.git",
    "directory": "packages/graphql-auth-transformer"
  },
  "author": "Amazon Web Services",
  "license": "Apache-2.0",
  "main": "lib/index.js",
  "types": "lib/index.d.ts",
  "keywords": [
    "graphql",
    "appsync",
    "aws"
  ],
  "scripts": {
    "test": "jest",
    "build": "tsc",
    "clean": "rimraf ./lib",
    "watch": "tsc -w",
    "extract-api": "ts-node ../../scripts/extract-api.ts"
  },
  "dependencies": {
    "graphql": "^15.5.0",
    "graphql-connection-transformer": "5.2.72",
    "graphql-mapping-template": "4.20.14",
    "graphql-transformer-common": "4.27.1",
    "graphql-transformer-core": "8.2.5"
  },
  "devDependencies": {
    "cloudform-types": "^4.2.0",
    "graphql-dynamodb-transformer": "7.2.72",
    "graphql-elasticsearch-transformer": "5.2.73",
    "graphql-function-transformer": "3.3.63",
    "lodash": "^4.17.21",
    "rimraf": "^3.0.0"
  },
  "jest": {
    "testURL": "http://localhost",
    "transform": {
      "^.+\\.tsx?$": "ts-jest"
    },
    "testRegex": "(src/__tests__/.*.test.ts)$",
    "moduleFileExtensions": [
      "ts",
      "tsx",
      "js",
      "jsx",
      "json",
      "node"
    ],
    "collectCoverage": true,
    "coverageProvider": "v8",
    "coverageThreshold": {
      "global": {
        "branches": 89,
        "functions": 90,
        "lines": 90
      }
    },
    "coverageReporters": [
      "clover",
      "text"
    ]
  }
}
