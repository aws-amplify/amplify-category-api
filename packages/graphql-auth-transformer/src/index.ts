export * from './ModelAuthTransformer';
// No-op change to trigger re-publish
