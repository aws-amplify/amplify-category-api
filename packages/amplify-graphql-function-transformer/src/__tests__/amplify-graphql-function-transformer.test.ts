import { Match, Template } from 'aws-cdk-lib/assertions';
import { parse } from 'graphql';
import { testTransform } from '@aws-amplify/graphql-transformer-test-utils';
import { AuthTransformer } from '@aws-amplify/graphql-auth-transformer';
import { FunctionTransformer } from '..';

test('for @function with only name, it generates the expected resources', () => {
  const validSchema = `
    type Query {
        echo(msg: String): String @function(name: "echofunction-\${env}")
    }
    `;

  const out = testTransform({
    schema: validSchema,
    transformers: [new FunctionTransformer()],
  });
  expect(out).toBeDefined();
  expect(out.stacks).toBeDefined();
  parse(out.schema);
  const stack = out.stacks.FunctionDirectiveStack;
  expect(stack).toBeDefined();
  Template.fromJSON(stack).resourceCountIs('AWS::IAM::Role', 1);
  Template.fromJSON(stack).resourceCountIs('AWS::IAM::Policy', 1);
  Template.fromJSON(stack).resourceCountIs('AWS::AppSync::DataSource', 1);
  Template.fromJSON(stack).resourceCountIs('AWS::AppSync::FunctionConfiguration', 1);
  Template.fromJSON(stack).resourceCountIs('AWS::AppSync::Resolver', 1);
  Template.fromJSON(stack).hasResourceProperties('AWS::IAM::Role', {
    AssumeRolePolicyDocument: {
      Statement: [
        {
          Action: 'sts:AssumeRole',
          Effect: 'Allow',
          Principal: {
            Service: 'appsync.amazonaws.com',
          },
        },
      ],
      Version: '2012-10-17',
    },
  });
  Template.fromJSON(stack).hasResourceProperties('AWS::IAM::Policy', {
    PolicyDocument: {
      Statement: [
        {
          Action: 'lambda:InvokeFunction',
          Effect: 'Allow',
          Resource: [
            {
              'Fn::If': [
                'HasEnvironmentParameter',
                {
                  'Fn::Sub': [
                    'arn:aws:lambda:${AWS::Region}:${AWS::AccountId}:function:echofunction-${env}',
                    { env: { Ref: Match.anyValue() } },
                  ],
                },
                { 'Fn::Sub': 'arn:aws:lambda:${AWS::Region}:${AWS::AccountId}:function:echofunction' },
              ],
            },
            {
              'Fn::Join': [
                '',
                [
                  {
                    'Fn::If': [
                      'HasEnvironmentParameter',
                      {
                        'Fn::Sub': [
                          'arn:aws:lambda:${AWS::Region}:${AWS::AccountId}:function:echofunction-${env}',
                          { env: { Ref: Match.anyValue() } },
                        ],
                      },
                      { 'Fn::Sub': 'arn:aws:lambda:${AWS::Region}:${AWS::AccountId}:function:echofunction' },
                    ],
                  },
                  ':*',
                ],
              ],
            },
          ],
        },
      ],
      Version: '2012-10-17',
    },
    PolicyName: Match.anyValue(),
    Roles: [{ Ref: Match.anyValue() }],
  });
  Template.fromJSON(stack).hasResourceProperties('AWS::AppSync::DataSource', {
    ApiId: { Ref: Match.anyValue() },
    Name: 'EchofunctionLambdaDataSource',
    Type: 'AWS_LAMBDA',
    LambdaConfig: {
      LambdaFunctionArn: {
        'Fn::If': [
          'HasEnvironmentParameter',
          {
            'Fn::Sub': ['arn:aws:lambda:${AWS::Region}:${AWS::AccountId}:function:echofunction-${env}', { env: { Ref: Match.anyValue() } }],
          },
          { 'Fn::Sub': 'arn:aws:lambda:${AWS::Region}:${AWS::AccountId}:function:echofunction' },
        ],
      },
    },
    ServiceRoleArn: {
      'Fn::GetAtt': ['EchofunctionLambdaDataSourceServiceRole3BE2FA57', 'Arn'],
    },
  });
  Template.fromJSON(stack).hasResourceProperties('AWS::AppSync::FunctionConfiguration', {
    ApiId: { Ref: Match.anyValue() },
    DataSourceName: { 'Fn::GetAtt': [Match.anyValue(), 'Name'] },
    FunctionVersion: '2018-05-29',
    Name: 'InvokeEchofunctionLambdaDataSource',
    RequestMappingTemplateS3Location: {
      'Fn::Join': [
        '',
        ['s3://', { Ref: Match.anyValue() }, '/', { Ref: Match.anyValue() }, '/resolvers/InvokeEchofunctionLambdaDataSource.req.vtl'],
      ],
    },
    ResponseMappingTemplateS3Location: {
      'Fn::Join': [
        '',
        ['s3://', { Ref: Match.anyValue() }, '/', { Ref: Match.anyValue() }, '/resolvers/InvokeEchofunctionLambdaDataSource.res.vtl'],
      ],
    },
  });
  Template.fromJSON(stack).hasResourceProperties('AWS::AppSync::Resolver', {
    ApiId: { Ref: Match.anyValue() },
    FieldName: 'echo',
    TypeName: 'Query',
    Kind: 'PIPELINE',
    PipelineConfig: {
      Functions: [{ 'Fn::GetAtt': [Match.anyValue(), 'FunctionId'] }],
    },
    RequestMappingTemplate: Match.anyValue(),
    ResponseMappingTemplateS3Location: {
      'Fn::Join': ['', ['s3://', { Ref: Match.anyValue() }, '/', { Ref: Match.anyValue() }, '/resolvers/Query.echo.res.vtl']],
    },
  });
  expect(out.resolvers).toMatchSnapshot();
});

test('for @function with account ID, it generates the expected resources', () => {
  const validSchema = `
    type Query {
        echo(msg: String): String @function(name: "echofunction", accountId: "123123456456")
    }
    `;

  const out = testTransform({
    schema: validSchema,
    transformers: [new FunctionTransformer()],
  });
  expect(out).toBeDefined();
  expect(out.stacks).toBeDefined();
  parse(out.schema);
  const stack = out.stacks.FunctionDirectiveStack;
  expect(stack).toBeDefined();
  Template.fromJSON(stack).resourceCountIs('AWS::IAM::Role', 1);
  Template.fromJSON(stack).resourceCountIs('AWS::IAM::Policy', 1);
  Template.fromJSON(stack).resourceCountIs('AWS::AppSync::DataSource', 1);
  Template.fromJSON(stack).resourceCountIs('AWS::AppSync::FunctionConfiguration', 1);
  Template.fromJSON(stack).resourceCountIs('AWS::AppSync::Resolver', 1);
  Template.fromJSON(stack).hasResourceProperties('AWS::IAM::Role', {
    AssumeRolePolicyDocument: {
      Statement: [
        {
          Action: 'sts:AssumeRole',
          Effect: 'Allow',
          Principal: {
            Service: 'appsync.amazonaws.com',
          },
        },
      ],
      Version: '2012-10-17',
    },
  });
  Template.fromJSON(stack).hasResourceProperties('AWS::IAM::Policy', {
    PolicyDocument: {
      Statement: [
        {
          Action: 'lambda:InvokeFunction',
          Effect: 'Allow',
          Resource: [
            {
              'Fn::If': [
                'HasEnvironmentParameter',
                {
                  'Fn::Sub': ['arn:aws:lambda:${AWS::Region}:123123456456:function:echofunction', {}],
                },
                { 'Fn::Sub': 'arn:aws:lambda:${AWS::Region}:123123456456:function:echofunction' },
              ],
            },
            {
              'Fn::Join': [
                '',
                [
                  {
                    'Fn::If': [
                      'HasEnvironmentParameter',
                      {
                        'Fn::Sub': ['arn:aws:lambda:${AWS::Region}:123123456456:function:echofunction', {}],
                      },
                      { 'Fn::Sub': 'arn:aws:lambda:${AWS::Region}:123123456456:function:echofunction' },
                    ],
                  },
                  ':*',
                ],
              ],
            },
          ],
        },
      ],
      Version: '2012-10-17',
    },
    PolicyName: Match.anyValue(),
    Roles: [{ Ref: Match.anyValue() }],
  });
  Template.fromJSON(stack).hasResourceProperties('AWS::AppSync::DataSource', {
    ApiId: { Ref: Match.anyValue() },
    Name: 'Echofunction123123456456LambdaDataSource',
    Type: 'AWS_LAMBDA',
    LambdaConfig: {
      LambdaFunctionArn: {
        'Fn::If': [
          'HasEnvironmentParameter',
          {
            'Fn::Sub': ['arn:aws:lambda:${AWS::Region}:123123456456:function:echofunction', {}],
          },
          { 'Fn::Sub': 'arn:aws:lambda:${AWS::Region}:123123456456:function:echofunction' },
        ],
      },
    },
    ServiceRoleArn: {
      'Fn::GetAtt': ['Echofunction123123456456LambdaDataSourceServiceRole0B60B47E', 'Arn'],
    },
  });
  Template.fromJSON(stack).hasResourceProperties('AWS::AppSync::FunctionConfiguration', {
    ApiId: { Ref: Match.anyValue() },
    DataSourceName: { 'Fn::GetAtt': [Match.anyValue(), 'Name'] },
    FunctionVersion: '2018-05-29',
    Name: 'InvokeEchofunction123123456456LambdaDataSource',
    RequestMappingTemplateS3Location: {
      'Fn::Join': [
        '',
        [
          's3://',
          { Ref: Match.anyValue() },
          '/',
          { Ref: Match.anyValue() },
          '/resolvers/InvokeEchofunction123123456456LambdaDataSource.req.vtl',
        ],
      ],
    },
    ResponseMappingTemplateS3Location: {
      'Fn::Join': [
        '',
        [
          's3://',
          { Ref: Match.anyValue() },
          '/',
          { Ref: Match.anyValue() },
          '/resolvers/InvokeEchofunction123123456456LambdaDataSource.res.vtl',
        ],
      ],
    },
  });
  Template.fromJSON(stack).hasResourceProperties('AWS::AppSync::Resolver', {
    ApiId: { Ref: Match.anyValue() },
    FieldName: 'echo',
    TypeName: 'Query',
    Kind: 'PIPELINE',
    PipelineConfig: {
      Functions: [{ 'Fn::GetAtt': [Match.anyValue(), 'FunctionId'] }],
    },
    RequestMappingTemplate: Match.anyValue(),
    ResponseMappingTemplateS3Location: {
      'Fn::Join': ['', ['s3://', { Ref: Match.anyValue() }, '/', { Ref: Match.anyValue() }, '/resolvers/Query.echo.res.vtl']],
    },
  });
  expect(out.resolvers).toMatchSnapshot();
});

test('two @function directives for the same lambda should produce a single datasource, single role and two resolvers', () => {
  const validSchema = `
    type Query {
        echo(msg: String): String @function(name: "echofunction-\${env}")
        magic(msg: String): String @function(name: "echofunction-\${env}")
    }
    `;

  const out = testTransform({
    schema: validSchema,
    transformers: [new FunctionTransformer()],
  });
  expect(out).toBeDefined();
  parse(out.schema);
  expect(out.stacks).toBeDefined();
  const stack = out.stacks.FunctionDirectiveStack;
  expect(stack).toBeDefined();
  Template.fromJSON(stack).resourceCountIs('AWS::IAM::Role', 1);
  Template.fromJSON(stack).resourceCountIs('AWS::IAM::Policy', 1);
  Template.fromJSON(stack).resourceCountIs('AWS::AppSync::DataSource', 1);
  Template.fromJSON(stack).resourceCountIs('AWS::AppSync::FunctionConfiguration', 1);
  Template.fromJSON(stack).resourceCountIs('AWS::AppSync::Resolver', 2);
});

test('two @function directives for the same field should be valid', () => {
  const validSchema = `
    type Query {
        echo(msg: String): String @function(name: "echofunction-\${env}") @function(name: "otherfunction")
    }
    `;

  const out = testTransform({
    schema: validSchema,
    transformers: [new FunctionTransformer()],
  });
  expect(out).toBeDefined();
  parse(out.schema);
  expect(out.stacks).toBeDefined();
  const stack = out.stacks.FunctionDirectiveStack;
  expect(stack).toBeDefined();
  Template.fromJSON(stack).resourceCountIs('AWS::AppSync::Resolver', 1);
  Template.fromJSON(stack).hasResourceProperties('AWS::AppSync::Resolver', {
    ApiId: { Ref: Match.anyValue() },
    FieldName: 'echo',
    TypeName: 'Query',
    Kind: 'PIPELINE',
    PipelineConfig: {
      Functions: [{ 'Fn::GetAtt': [Match.anyValue(), 'FunctionId'] }, { 'Fn::GetAtt': [Match.anyValue(), 'FunctionId'] }],
    },
  });
});

test('@function directive applied to Object should throw Error', () => {
  const invalidSchema = `
    type Query @function(name: "echofunction-\${env}") {
        echo(msg: String): String @function(name: "echofunction-\${env}")
    }
    `;

  expect(() =>
    testTransform({
      schema: invalidSchema,
      transformers: [new FunctionTransformer()],
    }),
  ).toThrow('Directive "@function" may not be used on OBJECT.');
});

test('includes auth info in stash', () => {
  const validSchema = `
    type Query {
      myFunction(userId: ID!): String
        @function(name: "myFunc-\${env}")
        @auth(rules: [{ allow: private, provider: iam }])
    }
    `;

  const out = testTransform({
    schema: validSchema,
    transformers: [new AuthTransformer(), new FunctionTransformer()],
    synthParameters: { identityPoolId: 'fake-test-id', adminRoles: ['fake-test-role'] },
    authConfig: {
      defaultAuthentication: {
        authenticationType: 'AWS_IAM',
      },
      additionalAuthenticationProviders: [],
    },
  });
  expect(out.stacks.FunctionDirectiveStack.Resources!.QuerymyFunctionResolver.Properties.RequestMappingTemplate).toMatchSnapshot();
});
