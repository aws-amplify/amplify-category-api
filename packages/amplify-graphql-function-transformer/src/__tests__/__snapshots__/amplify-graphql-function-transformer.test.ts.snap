// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`for @function with account ID, it generates the expected resources 1`] = `
Object {
  "InvokeEchofunction123123456456LambdaDataSource.req.vtl": "## [Start] Invoke AWS Lambda data source: Echofunction123123456456LambdaDataSource. **
{
  \\"version\\": \\"2018-05-29\\",
  \\"operation\\": \\"Invoke\\",
  \\"payload\\": {
      \\"typeName\\": $util.toJson($ctx.stash.get(\\"typeName\\")),
      \\"fieldName\\": $util.toJson($ctx.stash.get(\\"fieldName\\")),
      \\"arguments\\": $util.toJson($ctx.arguments),
      \\"identity\\": $util.toJson($ctx.identity),
      \\"source\\": $util.toJson($ctx.source),
      \\"request\\": $util.toJson($ctx.request),
      \\"prev\\": $util.toJson($ctx.prev)
  }
}
## [End] Invoke AWS Lambda data source: Echofunction123123456456LambdaDataSource. **",
  "InvokeEchofunction123123456456LambdaDataSource.res.vtl": "## [Start] Handle error or return result. **
#if( $ctx.error )
  $util.error($ctx.error.message, $ctx.error.type)
#end
$util.toJson($ctx.result)
## [End] Handle error or return result. **",
  "Query.echo.res.vtl": "$util.toJson($ctx.prev.result)",
}
`;

exports[`for @function with only name, it generates the expected resources 1`] = `
Object {
  "InvokeEchofunctionLambdaDataSource.req.vtl": "## [Start] Invoke AWS Lambda data source: EchofunctionLambdaDataSource. **
{
  \\"version\\": \\"2018-05-29\\",
  \\"operation\\": \\"Invoke\\",
  \\"payload\\": {
      \\"typeName\\": $util.toJson($ctx.stash.get(\\"typeName\\")),
      \\"fieldName\\": $util.toJson($ctx.stash.get(\\"fieldName\\")),
      \\"arguments\\": $util.toJson($ctx.arguments),
      \\"identity\\": $util.toJson($ctx.identity),
      \\"source\\": $util.toJson($ctx.source),
      \\"request\\": $util.toJson($ctx.request),
      \\"prev\\": $util.toJson($ctx.prev)
  }
}
## [End] Invoke AWS Lambda data source: EchofunctionLambdaDataSource. **",
  "InvokeEchofunctionLambdaDataSource.res.vtl": "## [Start] Handle error or return result. **
#if( $ctx.error )
  $util.error($ctx.error.message, $ctx.error.type)
#end
$util.toJson($ctx.result)
## [End] Handle error or return result. **",
  "Query.echo.res.vtl": "$util.toJson($ctx.prev.result)",
}
`;

exports[`includes auth info in stash 1`] = `
Object {
  "Fn::Join": Array [
    "",
    Array [
      "## [Start] Stash resolver specific context.. **
$util.qr($ctx.stash.put(\\"typeName\\", \\"Query\\"))
$util.qr($ctx.stash.put(\\"fieldName\\", \\"myFunction\\"))
$util.qr($ctx.stash.put(\\"authRole\\", \\"arn:aws:sts::",
      Object {
        "Ref": "AWS::AccountId",
      },
      ":assumed-role/",
      Object {
        "Ref": "referencetotransformerrootstackauthRoleNameFB872D50Ref",
      },
      "/CognitoIdentityCredentials\\"))
$util.qr($ctx.stash.put(\\"unauthRole\\", \\"arn:aws:sts::",
      Object {
        "Ref": "AWS::AccountId",
      },
      ":assumed-role/",
      Object {
        "Ref": "referencetotransformerrootstackunauthRoleName49F3C1FERef",
      },
      "/CognitoIdentityCredentials\\"))
$util.qr($ctx.stash.put(\\"identityPoolId\\", \\"fake-test-id\\"))
$util.qr($ctx.stash.put(\\"adminRoles\\", [\\"fake-test-role\\"]))
{}
## [End] Stash resolver specific context.. **",
    ],
  ],
}
`;
